lib/middlewares/src/metering.rs

```diff
@@ -0,0 +1,160 @@
+//! `metering` is a middleware for tracking how many operators are executed in total
+//! and putting a limit on the total number of operators executed.
+
+use std::fmt;
+use std::sync::Mutex;
+use wasmer::wasmparser::{
+    Operator, Result as WpResult, Type as WpType, TypeOrFuncType as WpTypeOrFuncType,
+};
+use wasmer::{
+    FunctionMiddleware, GlobalInit, GlobalType, LocalFunctionIndex, MiddlewareReaderState,
+    ModuleMiddleware, Mutability, Type,
+};
+use wasmer_types::GlobalIndex;
+use wasmer_vm::ModuleInfo;
+
+/// The module-level metering middleware.
+///
+/// # Panic
+///
+/// An instance of `Metering` should not be shared among different modules, since it tracks
+/// module-specific information like the global index to store metering state. Attempts to use
+/// a `Metering` instance from multiple modules will result in a panic.
```

Comment:
If an instance of `Metering` can be used for one module only, and it is installed via `CompilerConfig::push_middleware`, does this mean you need a new compiler (and engine and store) for each metered module?

Reply:
That's right. The current structure makes easier reusing the modules in headless engines.